### PR TITLE
Calendar state

### PR DESCRIPTION
**Pull request by ddomb**

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -4,6 +4,7 @@
     "@babel/preset-react"
   ],
   "plugins": [
+    "emotion",
```
Comment (IamShobe) [inline .babelrc]: move to last

File: app/frontend/src/components/Calendar/components/Cells.jsx [hunk 2]
```diff
@@ -1,138 +1,239 @@
...
+          {
+            eachDayOfInterval({
+              start: date,
+              end: endOfWeek(date)
+            }).map(date => DayRender(dateList,dates , date, monthStart, onDateClick))
+          }
+        </Week>
+      );
+    });
+  }, [currentDate, datesLoading]);
```
Comment (IamShobe) [inline app/frontend/src/components/Calendar/components/Cells.jsx]: separate to new component

File: app/frontend/src/components/Calendar/components/Cells.jsx [hunk 3]
```diff
@@ -1,138 +1,239 @@
...
+  status = fetchDates[datesInMonth-1].status;
+  if (fetchDates[datesInMonth-2].status === status) {
+    when = "End";
+  } else {
+    when = "Single";
+  }
+  return {status, when}; 
+}
+
+export const datesformatter = (fetchDates,today) => {
```
Comment (IamShobe) [inline app/frontend/src/components/Calendar/components/Cells.jsx]: dates

File: app/frontend/src/components/Calendar/components/Cells.jsx [hunk 4]
```diff
@@ -1,138 +1,239 @@
...
+const endWhen = (fetchDates, datesInMonth) => {
+  let status, when;
+  status = fetchDates[datesInMonth-1].status;
+  if (fetchDates[datesInMonth-2].status === status) {
+    when = "End";
+  } else {
+    when = "Single";
+  }
+  return {status, when}; 
+}
```
Comment (IamShobe) [inline app/frontend/src/components/Calendar/components/Cells.jsx]: add util function:
```
const iteratePrevCurrentNext = (iterator, callback) => {
    const toRet = [];
    const len = iterator.length;
    for(const i=0; i < len; i++){
        if (i === 0) toRet.push(callback(null, iterator[i], iterator[i + 1]));
        if (i === len - 1) toRet.push(callback(iterator[i - 1], iterator[i], null));
        toRet.push(callback(iterator[i - 1], iterator[i], iterator[i + 1]));
    }
    return toRet;
}
```
then usage
```
const datesList = iteratePrevCurrentNext(fetchDates, (prev, current, next) => {
    const dateObject = {status: current.status, when: "single"};
    if ((!prev || prev.status !== current.status) && (next && next.status === current.status)){
            dateObject.when = "start";
    } else if ((prev && prev.status === current.status) && (next && next.status === current.status)) {
            dateObject.when = "mid";
    } else if ((prev && prev.status === current.status) && (!next || next.status !== current.status))
       dateObject.when = "end";
    }
   return dateObject;
});
```

File: app/frontend/src/components/Calendar/components/Cells.jsx [hunk 5]
```diff
@@ -1,138 +1,239 @@
...
+    const dateStartDate = startOfWeek(dateMonthStart);
+    const dateEndDate = endOfWeek(addWeeks(dateMonthEnd, 1));
+    const dateToday  = new Date().getDate();
+    return {monthStart: dateMonthStart,
+            monthEnd: dateMonthEnd,
+            startDate: dateStartDate,
+            endDate: dateEndDate,
+            today: dateToday
+          };
+  })
```
Comment (IamShobe) [inline app/frontend/src/components/Calendar/components/Cells.jsx]: add a list of dependencies

File: app/frontend/src/components/Calendar/components/Cells.jsx [hunk 6]
```diff
@@ -1,138 +1,239 @@
...
+  const fetchParams = useMemo(() => {
+    return (
+      {start: getUnixTime(monthStart),
+      end: getUnixTime(monthEnd),
+      userId: userIdList}
+      );
+  });
```
Comment (IamShobe) [inline app/frontend/src/components/Calendar/components/Cells.jsx]: dependencies list

File: app/frontend/src/reducers/calendar.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+};
+
+
+export const calendarReducer = (state = initialState, action) => {
+  switch (action.type) {
+    case UPDATE_DATES:
+      return {
+        ...state,
+        dates: action.dates,
+        datesLoading: false,
```
Comment (IamShobe) [inline app/frontend/src/reducers/calendar.js]: just loading..
remember redux toolbelt?
see how they did async actions there.
perfect example :)

File: app/frontend/src/reducers/calendar.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
+import {UPDATE_DATES} from '~/actions/calendar';
+
+
+const initialState = {
+  dates: [],
```
Comment (IamShobe) [inline app/frontend/src/reducers/calendar.js]: maybe instead allow caching...
do it like so:
```
{
    data: {
       "01/01/2020": [...here the data for the date]
       "02/01/2020": ...
       .... and so on
    },
    loading=true
```

File: app/frontend/src/reducers/general.js [hunk 9]
```diff
@@ -21,11 +22,12 @@ export const generalReducer = (state = initialState, action) => {
...
-      return {
-        ...state,
-        dates: action.dates,
-      };
+    // case UPDATE_DATES:
+    //   return {
+    //     ...state,
+    //     dates: action.dates,
+    //     loading: false,
+    //   };
```
Comment (IamShobe) [inline app/frontend/src/reducers/general.js]: delete

File: app/frontend/src/reducers/general.js [hunk 10]
```diff
@@ -5,7 +5,8 @@ import AuthService from '~/services/auth';
 const initialState = {
   reasons: [],
   login: AuthService.isLoggedIn(),
-  dates: [],
+  // dates: [],
+  // loading:true
```
Comment (IamShobe) [inline app/frontend/src/reducers/general.js]: delete

File: app/frontend/src/services/date_datas.js [hunk 11]
```diff
@@ -38,19 +40,20 @@ class DateStatusService {
...
+      //         start: start,
+      //         end: end,
+      //         user_id: userId,
+      //       },
+      //       headers: {
+      //         ...AuthService.getAuthHeader(),
+      //       },
+      //     },
+      // );
+      // return response.data;
```
Comment (IamShobe) [inline app/frontend/src/services/date_datas.js]: add todo

File: tests/frontend/screens.test.js [hunk 12]
```diff
@@ -24,6 +24,7 @@ describe('Render all screens', () => {
     const {container} = renderWithRedux(<StyledApp/>, {
       history,
       store: createStore(() => ({
+        calendar: {datesLoading:true},
```
Comment (IamShobe) [inline tests/frontend/screens.test.js]: add a case where there is actuall data here and loading is false.
then coverage will bump to the sky :)

File: .babelrc (hunk 1 above)
Comment (IamShobe) [reply .babelrc]: NVM now when i know what it is (compatitor of styled componenets) get rid of it...
find other lib for loading like MUI or something else

Comment (ddomb) [reply .babelrc]: LOL, are U serious?

Comment (IamShobe) [reply .babelrc]: Yep..
We don't need conflicts of same libs

Comment (ddomb) [reply .babelrc]: .

File: app/frontend/src/components/Calendar/components/Cells.jsx (hunk 3 above)
Comment (ddomb) [reply app/frontend/src/components/Calendar/components/Cells.jsx]: .

File: app/frontend/src/components/Calendar/components/Cells.jsx (hunk 4 above)
Comment (ddomb) [reply app/frontend/src/components/Calendar/components/Cells.jsx]: .

File: app/frontend/src/components/Calendar/components/Cells.jsx (hunk 5 above)
Comment (ddomb) [reply app/frontend/src/components/Calendar/components/Cells.jsx]: .

File: app/frontend/src/components/Calendar/components/Cells.jsx (hunk 6 above)
Comment (ddomb) [reply app/frontend/src/components/Calendar/components/Cells.jsx]: .

File: app/frontend/src/reducers/calendar.js (hunk 7 above)
Comment (ddomb) [reply app/frontend/src/reducers/calendar.js]: .

File: app/frontend/src/reducers/general.js (hunk 9 above)
Comment (ddomb) [reply app/frontend/src/reducers/general.js]: .

File: app/frontend/src/reducers/general.js (hunk 10 above)
Comment (ddomb) [reply app/frontend/src/reducers/general.js]: .

File: app/frontend/src/services/date_datas.js (hunk 11 above)
Comment (ddomb) [reply app/frontend/src/services/date_datas.js]: .

**Pull request merged**